### PR TITLE
feat: update need newer file

### PR DESCRIPTION
**Pull request by SettingDust**
When targeting multi loader mods. Some mod has older version that mark as supporting prefer loader sorted by #36 will become the first.
This PR make the update require the files on particular platform newer than the local by recording the publishing date for project files.

### CODE REVIEW

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt [hunk 1]
```diff
@@ -34,29 +35,36 @@ suspend fun updateMultipleProjectsWithFiles(
...
+                }?.also { accProject ->
                     // Combine projects
-                    (accProject + newProject).get()?.let x@ { combinedProject ->
-                        if (combinedProject.hasNoFiles()) return@x // Do not update project if files are missing
-                        acc -= accProject
-                        acc += combinedProject
-                    }
+                    val accFiles = projectsToFiles[accProject]!!
+                    val accPublished = accFiles.find { it.type == platform.serialName }?.dataPublished
+                    if (accPublished != null && accPublished != Instant.MIN)
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt:52]: This probably is not backwards compatible. Old lock files don't have the "date published".

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt [hunk 2]
```diff
@@ -34,29 +35,36 @@ suspend fun updateMultipleProjectsWithFiles(
             }
     }
 
-    return@coroutineScope platforms.fold(projects.map { it.copy(files = mutableSetOf()) }.toMutableSet()) { acc, platform ->
+    val projectsToFiles = projects.associateTo(mutableMapOf()) { it.copy(files = mutableSetOf()) to it.files }
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt:38]: Why is this needed? Can you do it without this mutable map?

File: src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt [hunk 3]
```diff
@@ -27,6 +29,7 @@ data class ProjectFile(
     val hashes: MutableMap<String, String>? = null,
     @SerialName("required_dependencies") val requiredDependencies: MutableSet<String>? = null,
     val size: Int = 0,
+    val dataPublished: @Serializable(with = InstantIso8601Serializer::class) Instant = Instant.MIN,
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt:32]: This is not backwards compatible. It should be nullable and null by default.

**Review by juraj-hrivnak — CHANGES_REQUESTED** [review]

File: src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt (hunk 3 above)
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt:32]: It is backwards compatible, since it has a default value. ok

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt (hunk 1 above)
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt:52]: It will be `Instant.MIN` since there is a default value assigned

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt (hunk 2 above)
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt:38]: It's for finding the particular files from combined project which is different from the project in `projects`

File: src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt (hunk 3 above)
Comment (juraj-hrivnak) [reply src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt:32]: Ah, sorry I misread that.

Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt:32]: huh. Should I modify it to nullable?

Comment (juraj-hrivnak) [reply src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt:32]: No, I mean it was my mistake that I didn't notice the default `Instant.MIN` value,

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt (hunk 2 above)
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt:38]: It's combined projects. Not new I think.

Comment (juraj-hrivnak) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt:38]: Okay 👍

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt [hunk 4]
```diff
@@ -34,29 +35,36 @@ suspend fun updateMultipleProjectsWithFiles(
...
                     accProject.slug[platform.serialName] == newProject.slug[platform.serialName]
-                }?.let { accProject ->
+                }?.also { accProject ->
                     // Combine projects
-                    (accProject + newProject).get()?.let x@ { combinedProject ->
-                        if (combinedProject.hasNoFiles()) return@x // Do not update project if files are missing
-                        acc -= accProject
-                        acc += combinedProject
-                    }
+                    val accFiles = projectsToFiles[accProject]!!
```
[the remark this answers is not in the document]
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt:50]: The fold is iterating the keys of the map. And the lines below ensured the key exists.

[the remark this answers is not in the document]
Comment (juraj-hrivnak) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt]: Yes, but it shouldn't throw NPE when there is a bug. Maybe the function should return a `Result<MutableSet<Project>, ActionError>`

File: src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt [hunk 5]
```diff
@@ -27,6 +29,7 @@ data class ProjectFile(
     val hashes: MutableMap<String, String>? = null,
     @SerialName("required_dependencies") val requiredDependencies: MutableSet<String>? = null,
     val size: Int = 0,
+    @SerialName("data_published") val dataPublished: @Serializable(with = InstantIso8601Serializer::class) Instant = Instant.MIN,
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt:32]: There is a typo: `@SerialName("data_published") val dataPublished`
it should be: `@SerialName("date_published") val datePublished`

**Review by juraj-hrivnak — CHANGES_REQUESTED** [review]

File: src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (SettingDust) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt]: How to describe the error?

[the remark this answers is not in the document]
Comment (juraj-hrivnak) [reply src/commonMain/kotlin/teksturepako/pakku/api/actions/update/Update.kt]: `UpdateFailed(val project: Project, val reason: String) : ActionError("Failed to update ${project.type} ${project.slug}; $reason")`

For the message use `${dim(project.type)} ${project.getFlavoredSlug()}`
Also, the reason should tell what happened wrong.

File: src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt [hunk 6]
```diff
@@ -27,6 +29,7 @@ data class ProjectFile(
     val hashes: MutableMap<String, String>? = null,
     @SerialName("required_dependencies") val requiredDependencies: MutableSet<String>? = null,
     val size: Int = 0,
+    @SerialName("date_published") val dataPublished: @Serializable(with = InstantIso8601Serializer::class) Instant = Instant.MIN,
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/api/projects/ProjectFile.kt:32]: There is still a typo in `dataPublished`

File: src/commonMain/kotlin/teksturepako/pakku/cli/cmd/Update.kt [hunk 7]
```diff
@@ -44,7 +46,11 @@ class Update : CliktCommand()
 
         val updatedProjects = updateMultipleProjectsWithFiles(
             lockFile.getMcVersions(), lockFile.getLoaders(), currentProjects.toMutableSet(), ConfigFile.readOrNull(), numberOfFiles = 1
-        )
+        ).getOrElse {
+            terminal.danger(it.message())
```
Comment (juraj-hrivnak) [inline src/commonMain/kotlin/teksturepako/pakku/cli/cmd/Update.kt:50]: It should use `terminal.pError`

**Review by juraj-hrivnak — CHANGES_REQUESTED** [review]
